keps/369-job-interface/README.md

```diff
@@ -0,0 +1,381 @@
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+We collected feedbacks from the community about how to fully integrate Kueue with MPIJob,
+see [#499](https://github.com/kubernetes-sigs/kueue/issues/499).
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- Job interface defined here is a hint for developers to build their own controllers,
+it's not a hard constraint(to implement the interface) but a good reference.
+- This will increase the code complexity by wrapping the original Jobs.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Part I: Job Interface
+
+We will define a new interface named GenericJob, this should be implemented by custom job-like applications:
+
+```golang
+type GenericJob interface {
+  // Suspend instructs whether the job is suspended or not.
+  Suspend() bool
+  // Start will arm the job ready to start, like unsuspending the job and injecting nodeAffinity.
+  Start(ctx context.Context, wl *kueue.Workload) error
+  // Stop will disarm the job ready to stop, like suspending the job and restoring nodeAffinity.
+  Stop(ctx context.Context, wl *kueue.Workload) error
+  // InjectNodeAffinity will inject the node affinity extracting from workload to job.
+  InjectNodeAffinity(ctx context.Context, nodeSelectors []map[string]string) error
+  // RestoreNodeAffinity will restore the original node affinity of job.
+  RestoreNodeAffinity(ctx context.Context, nodeSelectors []map[string]string) error
+  // Finished means whether the job is completed/failed or not.
+  Finished() bool
+  // ConstructWorkload will build a workload corresponding to the job.
+  ConstructWorkload() (*kueue.Workload, error)
+  // PriorityClass returns the job's priority class name.
+  PriorityClass() string
+  // QueueName returns the queue name the job enqueued.
+  QueueName() string
+  // Ignored instructs whether this job should be ignored in reconciling, e.g. lacking the queueName.
+  Ignored() bool
+  // PodsCount returns the pods number job requires.
+  PodsCount() int32
```

Comment:
Should this return an array? Why do we need this in addition to ConstructWorkload?

Reply:
ConstructWorkload removed now.

Reply:
Yes, pushed.